src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c

```diff
@@ -0,0 +1,1557 @@
+ *          -   Before attaching the uprobes, set the version of the library to the
+ *              versions map. This module is version-dependent, and can only trace the library
+ *              of known versions.
+ *          -   Before using the module, initialize its percpu variables (used as "heap").
+ *              1.  grpc_c_event_data_local
+ *              2.  grpc_c_header_event_data_local
+ *              2.  grpc_c_metadata_local
+ */
+
+#include <linux/sched.h>
+#include "src/stirling/source_connectors/socket_tracer/bcc_bpf_intf/grpc_c.h"
+#include "src/stirling/bpf_tools/bcc_bpf/task_struct_utils.h"
+
+BPF_PERF_OUTPUT(grpc_c_events);
+BPF_PERF_OUTPUT(grpc_c_header_events);
+BPF_PERF_OUTPUT(grpc_c_close_events);
+
+struct list_pop_writable_stream_arguments
+{
+    void * transport;
+    void ** stream;
+};
+
+BPF_PERCPU_ARRAY(list_pop_writable_stream_arguments_local, struct list_pop_writable_stream_arguments, 1);
+BPF_PERCPU_ARRAY(grpc_c_metadata_local, struct grpc_c_metadata_t, 1);
+BPF_PERCPU_ARRAY(grpc_c_header_event_data_local, struct grpc_c_header_event_data_t, 1);
+BPF_PERCPU_ARRAY(grpc_c_event_data_local, struct grpc_c_event_data_t, 1);
+
+// Maps process id to version (@see grpc_c_version_e).
+BPF_HASH(grpc_c_versions, uint32_t, uint64_t, GRPC_C_DEFAULT_MAP_SIZE);
+
+// Size of a single slice is 0x20.
+// The refcount pointer is 0x8 bytes, then the union takes the amount of bytes needed for the larger option.
+// The larger option is the inlined slice, for which 0x18 bytes are needed.
+#define GRPC_SLICE_SIZE (0x20)
+
+// Types of structs in the grpc-c library.
+typedef void grpc_slice_buffer;
+typedef void grpc_slice;
+typedef void grpc_chttp2_transport;
+typedef void grpc_chttp2_stream;
+typedef void grpc_endpoint;
+typedef void grpc_metadata_batch;
+typedef void grpc_mdelem_list; // mdelem is short for metadata element
+typedef void grpc_linked_mdelem;
+typedef void grpc_mdelem_data;
+
+/*
+ * @brief   Initiate an empty grpc metadata struct.
+ *
+ * @return  NULL on failure - if user mode did not previously initiate our
+ *          heap.
+ *          Otherwise, a metadata struct on our percpu "heap".
+ */
+static inline struct grpc_c_metadata_t * initiate_empty_grpc_metadata()
+{
+    u32 zero = 0;
+
+    struct grpc_c_metadata_t * metadata = grpc_c_metadata_local.lookup(&zero);
+    if (NULL == metadata)
+    {
+        // User mode did not initiate the percpu buffer.
+        return NULL;
+    }
+
+#pragma unroll
+    for (u16 i = 0 ; i < MAXIMUM_AMOUNT_OF_ITEMS_IN_METADATA ; i++)
+    {
+        #pragma unroll
+        for (u16 j = 0 ; j < MAXIMUM_LENGTH_OF_VALUE_IN_METADATA ; j++)
+        {
+            ((volatile struct grpc_c_metadata_t *)metadata)->items[i].value[j] = 0;
+        }
+        #pragma unroll
+        for (u16 j = 0 ; j < MAXIMUM_LENGTH_OF_KEY_IN_METADATA ; j++)
+        {
+            ((volatile struct grpc_c_metadata_t *)metadata)->items[i].key[j] = 0;
+        }
+    }
+    metadata->count = 0;
```

Comment:
I agree, it's better when shorter. I changed the zero initialization functions to be much shorter, let me know what you think.

About the byte code count, I'm a little worried that if we don't initialize to zero at the start of the event, there will be events with outdated data, in case there's a flow where some field isn't updated, which might be terrible to debug (if we even notice it at all). How is this solved in pixie, do you just write exteremely-cautious code to make sure you always update everything?

Also, do you find this change important for runtime as well, or only for minimizing byte count for the 4.14 kernel? If it's only for minimizing byte count, perhaps we don't pass the restriction any way?